conda/core/solve.py

```diff
@@ -2153,6 +2155,155 @@ def working_state_reset(self):
+        self._logger = logging.getLogger(self.__class__.__name__)
+
+    def _set(self, key, value, *, reason=None, _level=3):
+        msg = f'{self._name}[{key}] set to `{value}`'
+        if reason:
+            msg += f' ({reason})'
+        self._logger.debug(msg, stacklevel=_level)
+        self._data[key] = value
+
+    @property
+    def name(self):
+        return self._name
+
+    def __iter__(self):
+        return iter(self._data)
+
+    def __len__(self):
+        return len(self._data)
+
+    def __getitem__(self, key):
+        return self._data[key]
+
+    def __setitem__(self, key, value):
+        self._set(key, value)
+
+    def __delitem__(self, key):
+        self._logger.debug(f'{self._name}[{key}] was deleted', stacklevel=2)
+        del self._data[key]
+
+    def pop(self, key):
+        self._logger.debug(f'{self._name}[{key}] was deleted', stacklevel=2)
+        return self._data.pop(key)
+
+    def popitem(self, key):
+        self._logger.debug(f'{self._name}[{key}] was deleted', stacklevel=2)
+        return self._data.popitem(key)
+
+    def clear(self):
+        self._logger.debug(f'{self._name} was cleared')
+        self._data.clear()
+
+    def update(self, data, **kwargs):
+        if data and kwargs:
+            raise ValueError('Can only update via a data object or keyword mapping')
+        self._logger.debug(f'{self._name}[{key}] was deleted', stacklevel=2)
+        return self._data.popitem(key)
+
+    def set(self, key, value, *, reason=None):
+        self._set(key, value)
+
+
+class SpecMap(Mapping):
+    def __init__(
+        self,
+        *,
+        frozen=None,
+        history=None,
+        neutered=None,
+    ):
+        # spec pools
+        self._frozen = TrackedMap(frozen or {})
+        self._history = TrackedMap(history or {})
+        self._neutered = TrackedMap(neutered or {})
+
+    @property
+    def _pools(self):
+        # ordered by priority
+        return (
+            self._frozen,
+            self._history,
+            self._neutered,
+        )
+
+    @functools.lru_cache
+    def _join_pools(self, *pools):
+        data = defaultdict(list)
+        for pool in pools:
+            for key, value in pool.items():
+                data[key].append(value)
+        return data
+
+    @property
+    def _data(self):
+        # we don't do this directly here to use lru_cache
+        return self._join_pools(self._pools)
+
+    def __getitem__(self, key):
+        return self._data[key]
+
+    def __iter__(self):
+        yield from self._data
+
+    def __len__(self):
+        return len(self._data)
+
+    def pool_items(self):
+        """Iterator that goes over the pool items one by one."""
+        for pool in self._pools:
+            for item in pool:
+                yield (pool, item)
+
+
+class SolverStateContainer:
+    def __init__(
+        self,
+        requested_specs_to_add: Mapping,
+        requested_specs_to_remove: Mapping,
+        prefix_data: Mapping,
```

Comment:
We only need to pass the `prefix` path here. Prefix data, history and pinnings can be calculated (one) from that.

Reply:
Good idea, the less we have to pass around the easier it'll be to evolve the code in the future.